x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/LogGroup.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.categorization;
+
+import org.apache.lucene.util.Accountable;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.RamUsageEstimator;
+import org.elasticsearch.core.Tuple;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.ml.aggs.categorization.CategorizationTokenTree.WILD_CARD;
+
+/**
+ * A log group that provides methods for:
+ *  - calculating similarity between it and a new log
+ *  - expanding the existing log group by adding a new log
+ */
+class LogGroup implements Accountable {
+
+    private final long id;
+    private final BytesRef[] logEvent;
+    private final long[] tokenCounts;
+    private long count;
+
+    // Used at the shard level for tracking the bucket ordinal for collecting sub aggregations
+    long bucketOrd;
+
+    @Override
+    public String toString() {
+        return "LogGroup{"
+            + "id="
+            + id
+            + ", logEvent="
+            + Arrays.stream(logEvent).map(BytesRef::utf8ToString).collect(Collectors.joining(", ", "[", "]"))
+            + ", count="
+            + count
+            + '}';
+    }
+
+    LogGroup(BytesRef[] logTokens, long count, long id) {
+        this.id = id;
+        this.logEvent = logTokens;
+        this.count = count;
+        this.tokenCounts = new long[logTokens.length];
+        Arrays.fill(this.tokenCounts, count);
+    }
+
+    public long getId() {
+        return id;
+    }
+
+    BytesRef[] getLogEvent() {
+        return logEvent;
+    }
+
+    public long getCount() {
+        return count;
+    }
+
+    Tuple<Double, Integer> calculateSimilarity(BytesRef[] logEvent) {
```

Comment:
I tend to avoid `Tuple` during aggregations just because its hard to know if the short lived tuple really gets stack allocated. I'm not sure of the right thing though because you wan to get it all in one pass.

Reply:
Yeah, I can see about refactoring this.